File: daily/2024/1937_Maximum_Number_of_Points_with_Cost.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    long long maxPoints(vector<vector<int>>& points) {
        int m = points.size();
        int n = points[0].size();
        vector<long long> dp(n);
        vector<long long> pre_dp(n);
        for (int j = 0; j < n; j++) pre_dp[j] = points[0][j];
        for (int i = 1; i < m; i++){
            long long temp_max = -1;
            for (int j = 0; j < n; j++){
                temp_max--;
                temp_max = max(temp_max, pre_dp[j]);
                dp[j] = temp_max + points[i][j];
            }
            temp_max = -1;
            for (int j = n - 1; j >= 0; j--){
                temp_max--;
                temp_max = max(temp_max, pre_dp[j]);
                dp[j] = max(dp[j], temp_max + points[i][j]);
            }
            swap(dp, pre_dp);
        }
        long long ans = 0;
        for (long long i : pre_dp) ans = max(ans, i);
        return ans;
    }
};

int main(){
    vector<vector<int>> points = {{1,2,3},{1,5,1},{3,1,1}};
    // vector<vector<int>> points = {{1,5},{2,3},{4,2}};
    Solution S;
    cout << S.maxPoints(points) << endl;
    return 0;
}

File: daily_python/1937_Maximum_Number_of_Points_with_Cost.py
import os
import sys
from typing import List

class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        row = len(points)
        col = len(points[0])
        pre_dp = points[0]
        dp = [0] * col
        for i in range(1, row):
            temp_max = 0
            for j in range(col):
                temp_max = max(temp_max - 1, pre_dp[j])
                dp[j] = points[i][j] + temp_max
            temp_max = 0
            for j in range(col-1, -1, -1):
                temp_max = max(temp_max - 1, pre_dp[j])
                dp[j] = max(dp[j], points[i][j] + temp_max)
            pre_dp = dp.copy()
        return max(pre_dp)
    
def readDataSet(filename: str)-> None:
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            points = lines[0].split('=')[1][1:-1].replace('{', '[').replace('}',']')
            points = eval(points)
            dataset.append(points)
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for points in dataset:
        results.append(solution.maxPoints(points))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')